syntax = "proto3";
package RevaProtocol;
option java_package = "reva.protocol";

service RevaDataService {
    rpc getDataAtAddress(RevaGetDataAtAddressRequest) returns (RevaGetDataAtAddressResponse) {}
    rpc getListData(RevaDataListRequest) returns (stream RevaDataListResponse) {}
    rpc getStringList(RevaStringListRequest) returns (stream RevaStringListResponse) {}
}

message RevaGetDataAtAddressRequest {
    string address = 1;
    string symbol = 2;
    uint64 size = 3;
}

message RevaGetDataAtAddressResponse {
    string address = 1;
    bytes data = 2;
    string symbol = 3;
    string type = 4;
    repeated string incoming_references = 5;
    repeated string outgoing_references = 6;
    uint64 size = 7;
}

message RevaDataListRequest {
}

message RevaDataListResponse {
    string address = 1;
    string symbol = 2;
    string type = 3;
    uint64 size = 4;
    repeated string incoming_references = 5;
    repeated string outgoing_references = 6;
}

message RevaStringListRequest {
}

message RevaStringListResponse {
    string address = 1;
    string symbol = 2;
    string value = 3;
    repeated string incoming_references = 4;
    repeated string outgoing_references = 5;
}